### PR TITLE
csi: add NodeServer interface implementation

### PR DESCRIPTION
**Pull request by fatih**
This adds the NodeServer interface of the CSI spec. The test suite is passing and is also enabled in this PR. However there are many things todo yet and is not tested against a real K8S server.

### CODE REVIEW

File: do/csi/driver/mounter.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	if source == "" {
+		return errors.New("source is not specified for formatting the volume")
+	}
+
+	mkfsArgs = append(mkfsArgs, source)
+	if fsType == "ext4" || fsType == "ext3" {
+		mkfsArgs = []string{"-F", source}
+	}
+
+	out, err := exec.Command(mkfsCmd, mkfsArgs...).CombinedOutput()
```
Comment (lxfontes) [inline do/csi/driver/mounter.go]: check if `mkfs.%s` is present with https://golang.org/pkg/os/exec/#LookPath prior executing

Comment (fatih) [reply do/csi/driver/mounter.go]: good call, fixed with: https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/77/commits/f763ca9bf39d0a92f72512f3feca116de709b20b

**Comment by fatih** [comment]
I've decided to open other small PR's, so would like to merge this. PTAL again.

File: do/csi/driver/node.go [hunk 2]
```diff
@@ -64,3 +180,11 @@ func (d *Driver) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCapabi
 		},
 	}, nil
 }
+
+// getDiskSource returns the absolute path of the attached volume for the given
+// DO volume name
+func getDiskSource(volumeName string) string {
+	diskIDPath := "/dev/disk/by-id"
```
Comment (cagedmantis) [inline do/csi/driver/node.go]: Should these be const?

**Comment by lxfontes** [comment]
lgtm

File: do/csi/driver/driver_test.go [hunk 3]
```diff
@@ -79,19 +79,35 @@ func (f *fakeAPI) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			// a list of volumes for the given name
+			for _, vol := range f.volumes {
+				if vol.Name == name {
+					var resp = struct {
+						Volumes []*godo.Volume
+						Links   *godo.Links
+					}{
+						Volumes: []*godo.Volume{vol},
+					}
+					_ = json.NewEncoder(w).Encode(&resp)
```
Comment (cagedmantis) [inline do/csi/driver/driver_test.go]: handle error

File: do/csi/driver/driver_test.go [hunk 4]
```diff
@@ -79,19 +79,35 @@ func (f *fakeAPI) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			// single volume get
+			id := filepath.Base(r.URL.Path)
+			vol := f.volumes[id]
+			var resp = struct {
+				Volume *godo.Volume
+				Links  *godo.Links
+			}{
+				Volume: vol,
+			}
+			_ = json.NewEncoder(w).Encode(&resp)
```
Comment (cagedmantis) [inline do/csi/driver/driver_test.go]: Handle error.

File: do/csi/driver/node.go (hunk 2 above)
Comment (fatih) [reply do/csi/driver/node.go]: changed them const

File: do/csi/driver/driver_test.go (hunk 3 above)
Comment (fatih) [reply do/csi/driver/driver_test.go]: fixed

File: do/csi/driver/driver_test.go (hunk 4 above)
Comment (fatih) [reply do/csi/driver/driver_test.go]: fixed

**Comment by fatih** [comment]
Merging this, other follow up PR's will follow soon.

File: do/csi/driver/driver_test.go [hunk 5]
```diff
@@ -61,9 +55,15 @@ func TestDriverSuite(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	mntStageDir, err := ioutil.TempDir("", "mnt-stage")
```
Comment (cagedmantis) [inline do/csi/driver/driver_test.go]: Should we defer a directory delete

Comment (fatih) [reply do/csi/driver/driver_test.go]: @cagedmantis I'll do a follow up PR with this fix included.

File: do/csi/driver/mounter.go (hunk 1 above)
Comment (cagedmantis) [reply do/csi/driver/mounter.go]: nit. Might be a good idea to use the path returned by `LookPath` for the mkfsCmd. Not necessary

**Pull request merged**